src/storage/src/source/reclock.rs

```diff
@@ -47,6 +47,91 @@ pub struct ReclockFollower {
     inner: Rc<RefCell<ReclockFollowerInner>>,
 }
 
+/// Packs a binding into a Row.
+///
+/// A binding of None partition is encoded as a single datum containing the offset.
+///
+/// A binding of a Kafka partition is encoded as the partition datum followed by the offset datum.
+fn pack_binding(pid: PartitionId, offset: MzOffset) -> SourceData {
+    let mut row = Row::with_capacity(2);
+    let mut packer = row.packer();
+    match pid {
+        PartitionId::None => {}
+        PartitionId::Kafka(pid) => packer.push(Datum::Int32(pid)),
+    }
+    packer.push(Datum::UInt64(offset.offset));
+    SourceData(Ok(row))
+}
+
+/// Unpacks a binding from a Row
+/// See documentation of [pack_binding] for the encoded format
+fn unpack_binding(data: SourceData) -> (PartitionId, MzOffset) {
+    let row = data.0.expect("invalid binding");
+    let mut datums = row.iter();
+    let (pid, offset_datum) = match (datums.next(), datums.next()) {
+        (Some(Datum::Int32(pid)), Some(offset_datum)) => (PartitionId::Kafka(pid), offset_datum),
+        (Some(offset_datum), None) => (PartitionId::None, offset_datum),
+        _ => panic!("invalid binding"),
+    };
+
+    let offset = match offset_datum {
+        Datum::UInt64(offset) => MzOffset::from(offset),
+        _ => panic!("invalid binding"),
+    };
+    (pid, offset)
+}
+
+/// Drains the provided vector of updates containing insertions and retractions of full offset
+/// statements and differentiates them into updates containing the change in offset in the diff
+/// field.
+fn differentiate(
+    updates: &mut Vec<((PartitionId, MzOffset), Timestamp, Diff)>,
+) -> impl Iterator<Item = (PartitionId, Timestamp, MzOffset)> + '_ {
+    consolidation::consolidate_updates(updates);
```

Comment:
is this strictly necessary? or does it just reduce the size of the remap trace

Reply:
It is necessary for the updates to be sorted, but the consolidation will also reduce the work we need to do